File: backend/src/routes/projects.ts
import { Router, Request, Response } from "express";
import { supabase } from "../lib/supabase";
import { anomalyService, DEFAULT_THRESHOLDS } from "../services/anomaly";

const router = Router();

/**
 * @route   GET /api/projects
 * @desc    Return all projects
 */
router.get("/", async (req: Request, res: Response) => {
  try {
    const { userId } = req.query;
    
    let dbQuery = supabase
      .from("projects")
      .select("*")
      .order("created_at", { ascending: false });

    if (userId) {
      dbQuery = dbQuery.eq("user_id", userId);
    }

    const { data, error } = await dbQuery.limit(100);

    if (error) {
      console.error("Projects Fetch Error:", error);
      return res
        .status(500)
        .json({ error: "Failed to fetch projects", details: error.message });
    }

    return res.json({ projects: data || [] });
  } catch (error: any) {
    console.error("Projects Route Error:", error);
    return res
      .status(500)
      .json({ error: "Internal server error", details: error.message });
  }
});

/**
 * @route   GET /api/projects/:projectId
 * @desc    Get a single project
 */
router.get("/:projectId", async (req: Request, res: Response) => {
  try {
    const { projectId } = req.params;

    const { data, error } = await supabase
      .from("projects")
      .select("*")
      .eq("id", projectId)
      .single();

    if (error) {
      console.error("Project Fetch Error:", error);
      return res
        .status(500)
        .json({ error: "Failed to fetch project", details: error.message });
    }

    if (!data) {
      return res.status(404).json({ error: "Project not found" });
    }

    return res.json(data);
  } catch (error: any) {
    console.error("Project Route Error:", error);
    return res
      .status(500)
      .json({ error: "Internal server error", details: error.message });
  }
});

/**
 * @route   POST /api/projects
 * @desc    Create a new project
 */
router.post("/", async (req: Request, res: Response) => {
  try {
    const { name, user_id, thresholds } = req.body;

    if (!name) {
      return res.status(400).json({ error: "Project name is required" });
    }

    if (!user_id) {
      return res.status(400).json({ error: "user_id is required" });
    }

    const { data, error } = await supabase
      .from("projects")
      .insert([{ name, user_id, thresholds: normalizeThresholds(thresholds) }])
      .select()
      .single();

    if (error) {
      console.error("Project Creation Error:", error);
      return res
        .status(500)
        .json({ error: "Failed to create project", details: error.message });
    }

    return res.status(201).json(data);
  } catch (error: any) {
    console.error("Project Create Route Error:", error);
    return res
      .status(500)
      .json({ error: "Internal server error", details: error.message });
  }
});

/**
 * @route   GET /api/projects/:projectId/thresholds/recommendation
 * @desc    Recommend thresholds from recent metric history
 */
router.get("/:projectId/thresholds/recommendation", async (req: Request, res: Response) => {
  try {
    const { projectId } = req.params;
    const recommendation = await anomalyService.getRecommendedThresholds(projectId);

    return res.json(recommendation);
  } catch (error: any) {
    console.error("Threshold Recommendation Error:", error);
    return res
      .status(500)
      .json({ error: "Failed to recommend thresholds", details: error.message });
  }
});

/**
 * @route   PATCH /api/projects/:projectId/thresholds
 * @desc    Update per-project alert thresholds
 */
router.patch("/:projectId/thresholds", async (req: Request, res: Response) => {
  try {
    const { projectId } = req.params;
    const thresholds = normalizeThresholds(req.body.thresholds || req.body);

    const { data, error } = await supabase
      .from("projects")
      .update({ thresholds })
      .eq("id", projectId)
      .select()
      .single();

    if (error) {
      console.error("Threshold Update Error:", error);
      return res
        .status(500)
        .json({ error: "Failed to update thresholds", details: error.message });
    }

    return res.json(data);
  } catch (error: any) {
    console.error("Threshold Route Error:", error);
    return res
      .status(500)
      .json({ error: "Internal server error", details: error.message });
  }
});

function normalizeThresholds(input: any) {
  const merged = {
    ...DEFAULT_THRESHOLDS,
    ...(input || {}),
  };

  return {
    cpu: clampThreshold(merged.cpu, 1, 100, DEFAULT_THRESHOLDS.cpu),
    memory: clampThreshold(merged.memory, 1, 100, DEFAULT_THRESHOLDS.memory),
    requests: clampThreshold(merged.requests, 1, 100000, DEFAULT_THRESHOLDS.requests),
    errors: clampThreshold(merged.errors, 0, 100000, DEFAULT_THRESHOLDS.errors),
  };
}

function clampThreshold(value: any, min: number, max: number, fallback: number) {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) return fallback;
  return Math.min(max, Math.max(min, parsed));
}

export default router;

File: backend/app/routes/projects.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from uuid import uuid4
from app.database import get_supabase
from supabase import Client

router = APIRouter()

class ProjectCreate(BaseModel):
    name: str
    user_id: str

class ProjectResponse(BaseModel):
    id: str
    name: str
    user_id: str
    created_at: str
    updated_at: str

@router.post("/projects", response_model=ProjectResponse)
async def create_project(project: ProjectCreate, supabase: Client = Depends(get_supabase)):
    """Create a new monitoring project"""
    try:
        # Validate user_id is not empty
        if not project.user_id or not project.user_id.strip():
            raise HTTPException(status_code=400, detail="user_id is required")
        
        if not project.name or not project.name.strip():
            raise HTTPException(status_code=400, detail="Project name is required")
        
        project_id = str(uuid4())
        
        # Insert into Supabase
        result = supabase.table("projects").insert({
            "id": project_id,
            "name": project.name.strip(),
            "user_id": project.user_id,
        }).execute()
        
        if result.data and len(result.data) > 0:
            return result.data[0]
        else:
            raise HTTPException(status_code=500, detail="Failed to create project in database")
            
    except HTTPException:
        raise
    except Exception as e:
        print(f"Error creating project: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/projects")
async def list_projects(user_id: str = None, supabase: Client = Depends(get_supabase)):
    """Get all projects for a user"""
    try:
        if user_id:
            result = supabase.table("projects").select("*").eq("user_id", user_id).execute()
        else:
            result = supabase.table("projects").select("*").execute()
        
        return {"projects": result.data if result.data else []}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/projects/{project_id}", response_model=ProjectResponse)
async def get_project(project_id: str, supabase: Client = Depends(get_supabase)):
    """Get a specific project"""
    try:
        result = supabase.table("projects").select("*").eq("id", project_id).execute()
        
        if result.data and len(result.data) > 0:
            return result.data[0]
        else:
            raise HTTPException(status_code=404, detail="Project not found")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.put("/projects/{project_id}")
async def update_project(project_id: str, name: str, supabase: Client = Depends(get_supabase)):
    """Update a project"""
    try:
        if not name or not name.strip():
            raise HTTPException(status_code=400, detail="Project name is required")
        
        result = supabase.table("projects").update({
            "name": name.strip()
        }).eq("id", project_id).execute()
        
        if result.data and len(result.data) > 0:
            return result.data[0]
        else:
            raise HTTPException(status_code=404, detail="Project not found")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/projects/{project_id}")
async def delete_project(project_id: str, supabase: Client = Depends(get_supabase)):
    """Delete a project"""
    try:
        result = supabase.table("projects").delete().eq("id", project_id).execute()
        return {"message": "Project deleted successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
